service-coordtransform/src/main/java/fi/nls/paikkatietoikkuna/coordtransform/CoordinateTransformationActionHandler.java

```diff
@@ -83,20 +197,181 @@ public void handleAction(ActionParameters params) throws ActionException {
         } catch (IllegalArgumentException e) {
             throw new ActionException(e.getMessage());
         }
 
         HttpServletResponse response = params.getResponse();
-        response.setContentType(IOHelper.CONTENT_TYPE_JSON);
         try (OutputStream out = response.getOutputStream()) {
-            writeJsonResponse(out, coords, dimension);
+            if (transformToFile){
+                String fileName = addFileExt(exportSettings.getFileName());
+                writeFile (coords, targetDimension, exportSettings, targetCrs);
+                response.setContentType(FILE_TYPE);
+                response.setHeader("Content-Disposition", "attachment; filename=" + fileName);
+                response.setHeader("Content-Length", String.valueOf(new File(fileName).length()));
+                FileInputStream  fis = new FileInputStream(fileName);
+                IOUtils.copy(fis,out);
+                out.flush();
+            } else {
+                response.setContentType(IOHelper.CONTENT_TYPE_JSON);
+                writeJsonResponse(out, coords, inputCoords, targetDimension);
+            }
         } catch (IOException e) {
             throw new ActionException("Failed to write JSON to client");
         }
     }
+    private List<Coordinate> getCoordsFromJsonArray (ActionParameters params, int dimension, boolean addZeroes) throws ActionException{
+        List<Coordinate> coords;
+        try (InputStream in = params.getRequest().getInputStream()) {
+            coords = parseInputCoordinates(in, dimension, addZeroes);
+        } catch (IOException e) {
+            throw new ActionException("Failed to parse input JSON!");
+        }
+        return coords;
+    }
 
+    protected List<Coordinate> getCoordsFromFile (CoordTransFile sourceOptions, FileItem file, int dimension, boolean addZeroes, boolean storeLineEnds) throws ActionException{
+        List<Coordinate> coordinates = new ArrayList<>();
+        String line;
+        String[] coords;
+        int xIndex = 0;
+        int yIndex = 1;
+        int zIndex = 2;
+        int coordDimension = dimension;
+        int headerLineCount = sourceOptions.getHeaderLineCount();
+        String coordSeparator = coordinateSeparators.get(sourceOptions.getCoordinateSeparator());
+        if (sourceOptions.isAxisFlip()){
+            xIndex = 1;
+            yIndex = 0;
+        }
+        if (sourceOptions.isPrefixId()){
+            xIndex++;
+            yIndex++;
+            zIndex++;
+            coordDimension++;
+        }
+        boolean replaceCommas = false;
+        if (sourceOptions.getDecimalSeparator()==','){
+            replaceCommas = true;
+        }
+        String unit = sourceOptions.getUnit();
+        boolean transformUnit = false;
+        if (unit != null && !unit.equals(DEGREE)){
+            transformUnit = true;
+        }
+        double x,y,z;
+        try{
+            BufferedReader br = new BufferedReader(
```

Comment:
Use try-with-resources. `try (BufferedReader br = new BufferedReader(...) {`
Currently the Reader is left open in case an exception occurs.